Sources/Plasma/Apps/plClient/Mac-Cocoa/PLSPatcher.mm

```diff
@@ -106,6 +114,52 @@ - (void)start
     self.patcher->Start();
 }
 
+- (NSURL *)completeSelfPatch:(NSError **)error;
+{
+    NSString* destinationPath = [NSString stringWithSTString:plManifest::PatcherExecutable().AsString()];
+    NSURL* destinationURL = [NSURL fileURLWithPath:[NSString stringWithSTString:plManifest::PatcherExecutable().AsString()]];
+    
+    NSError* errorInScope;
+    
+    if (!self.updatedClientURL) {
+        // uh oh - this implies we weren't able to decompress the client
+        if (error)
+        {
+            // Handle as a generic could not read file error.
+            // Bad compression on the server will require correction on the server end.
+            *error = [NSError errorWithDomain:NSCocoaErrorDomain code:NSFileReadNoSuchFileError userInfo:nil];
+        }
+        return;
+    }
+    
+    if ([NSFileManager.defaultManager fileExistsAtPath:destinationPath]) {
+        // need to swap
+        BOOL swapSucceeded = renamex_np(destinationURL.path.fileSystemRepresentation, self.updatedClientURL.path.fileSystemRepresentation, RENAME_SWAP) == 0;
+        if (swapSucceeded) {
+            // delete the old version - this is very likely us
+            // we want to terminate after. Our bundle will no longer be valid.
+            [NSFileManager.defaultManager removeItemAtURL:self.updatedClientURL error:&errorInScope];
+        } else {
+            // abort and return an error
+            errorInScope = [NSError errorWithDomain:NSPOSIXErrorDomain code:errno userInfo:nil];
+        }
+    } else {
+        // no executable already present! Just move things into place.
+        [NSFileManager.defaultManager moveItemAtURL:self.updatedClientURL toURL:destinationURL error:&errorInScope];
+    }
+    
+    if (errorInScope) {
+        // Try to clean up if there was an error
+        [NSFileManager.defaultManager removeItemAtURL:self.updatedClientURL error:nil];
+        if (error) {
+            *error = errorInScope;
+        }
+        return nil;
+    }
+    
+    return destinationURL;
```

Comment:
This is specifically against Cocoa best practice:
https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ErrorHandlingCocoa/CreateCustomizeNSError/CreateCustomizeNSError.html#//apple_ref/doc/uid/TP40001806-CH204-BAJIIGCC

> Important: You should only modify the NSError parameter if there is an error in your method and your method directly returns NO. Always test to see if the parameter is non-NULL before assigning an object to it. And never assign nil to an error parameter.

Error in/outs should be allowed to chain between NSError calls so you should never overwrite an existing value unless there is an error to actually return. You could clear and hide an error from a previous call.

Reply:
Ah, I wasn't aware of that. Then we should follow that convention of course.

In that case, the `if (error)` condition needs to be changed though, because the `error` variable will be left uninitialized in the non-error case. You're already checking that `tempDirectory` is non-`nil`, but you also need to check the `BOOL` return value of [`-[NSFileManager createDirectoryAtURL:withIntermediateDirectories:attributes:error:]`](https://developer.apple.com/documentation/foundation/nsfilemanager/1415371-createdirectoryaturl).

Actually, perhaps we don't need the `error` out parameter after all, because the code doesn't do anything with the `error` value. For just checking success/failure, the return values are enough.

Reply:
> In that case, the if (error) condition needs to be changed though, because the error variable will be left uninitialized in the non-error case.

Obj-C errors are implicitly initialized to nil upon creation - so the variable will be initialized.

> You're already checking that tempDirectory is non-nil, but you also need to check the BOOL return value of -[NSFileManager createDirectoryAtURL:withIntermediateDirectories:attributes:error:].

In Obj-C error patterns - the state of the bool will always match the state of the error. I.E. if the error is nil then YES will be returned, if the error is not nil than NO will be returned. So it should be sufficient to check the error or the bool value, but it's not required to check both.

Reply:
> Obj-C variables are implicitly initialized to nil upon creation

Ah, apparently this guarantee was added with ARC - I wasn't aware of that. Yeah, in that case the code actually works fine. (I still think you can omit the `error` variable here for simplicity, but you're right that it won't change the behavior of the code.)

Reply:
Interesting. This code is meant to be ARC - but given that there has been ARC-less code in the code base, I'd feel much better initializing it by hand. I can make the change.

I'll need to make a mental note on that. I did not realize that feature came in only with ARC.